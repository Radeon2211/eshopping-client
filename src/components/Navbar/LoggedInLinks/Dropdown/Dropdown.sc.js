import styled from 'styled-components';
import { motion } from 'framer-motion';

// eslint-disable-next-line import/prefer-default-export
export const Wrapper = styled(motion.div)`
  background-color: ${({ theme }) => theme.colors.light1};
  border: 1px solid ${({ theme }) => theme.colors.light3};
  border-radius: 1px;
  overflow: hidden;
  position: absolute;
  top: calc(100% + ${({ theme }) => theme.spacings.level2});
  right: 0;
  width: 25rem;
  z-index: ${({ theme }) => theme.zIndexes.level2};

  & .list {
    list-style: none;
  }

  & .item {
    &:not(:last-child) {
      border-bottom: 1px solid ${({ theme }) => theme.colors.light2};
    }
  }

  & .link,
  & .link:link,
  & .link:visited {
    align-items: center;
    color: #000;
    display: flex;
    font-size: 1.4rem;
    justify-content: center;
    padding: ${({ theme }) => theme.spacings.level2};
    text-align: center;
    text-decoration: none;
    transition: color 0.12s;
    width: 100%;

    &:hover {
      color: ${({ theme }) => theme.colors.green};
    }

    & > *:not(:first-child) {
      margin-left: ${({ theme }) => theme.spacings.level1};
    }
  }

  @media only screen and (max-width: 37.5em) {
    & .link:link,
    & .link:visited {
      font-size: 1.5rem;
    }
  }
`;
