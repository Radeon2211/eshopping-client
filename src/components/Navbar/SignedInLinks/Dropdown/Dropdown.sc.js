import styled, { css } from 'styled-components';

// eslint-disable-next-line import/prefer-default-export
export const Wrapper = styled.div`
  background-color: ${({ theme }) => theme.colors.light1};
  border: 1px solid ${({ theme }) => theme.colors.light3};
  border-radius: 1px;
  opacity: 0;
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  top: calc(100% + ${({ theme }) => theme.spacings.level2});
  transform: translateY(-10%);
  transition: all 0.2s;
  right: 0;
  width: 25rem;
  z-index: ${({ theme }) => theme.zIndexes.level2};

  ${({ visible }) =>
    visible &&
    css`
      opacity: 1;
      pointer-events: initial;
      transform: translateY(0);
    `}

  & .list {
    list-style: none;
  }

  & .item {
    &:not(:last-child) {
      border-bottom: 1px solid ${({ theme }) => theme.colors.light2};
    }
  }

  & .link:link,
  & .link:visited {
    align-items: center;
    color: #000;
    display: flex;
    font-size: 1.4rem;
    justify-content: center;
    padding: ${({ theme }) => theme.spacings.level2};
    text-align: center;
    text-decoration: none;
    transition: color 0.1s;
    width: 100%;

    &:hover {
      color: ${({ theme }) => theme.colors.green};
    }

    & > *:not(:first-child) {
      margin-left: ${({ theme }) => theme.spacings.level1};
    }
  }

  @media only screen and (max-width: 37.5em) {
    & .link:link,
    & .link:visited {
      font-size: 1.5rem;
    }
  }
`;
