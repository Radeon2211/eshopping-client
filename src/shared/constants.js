import * as Yup from 'yup';
import { getCountries } from 'country-fns';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productConditions = {
  NEW: 'new',
  USED: 'used',
  NOT_APPLICABLE: 'not_applicable',
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([productConditions.NEW, productConditions.USED, productConditions.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
};

export const listOfAreaCodes = getCountries().map(({ name, dial }) => {
  const finalValue = `+${dial} ${name.split('(')[0].trim()}`;
  return {
    value: dial,
    label: finalValue,
  };
});

export const listOfCountries = getCountries().map(({ name }) => {
  const finalValue = name.split('(')[0].trim();
  return {
    value: finalValue,
    label: finalValue,
  };
});

export const productPages = {
  ALL_PRODUCTS: 'ALL_PRODUCTS',
  MY_PRODUCTS: 'MY_PRODUCTS',
  USER_PRODUCTS: 'USER_PRODUCTS',
};

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const orderTypes = {
  PLACED_ORDERS: 'PLACED_ORDERS',
  SELL_HISTORY: 'SELL_HISTORY',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const modalTypes = {
  ABOUT_WEBSITE: 'ABOUT_WEBSITE',
  ADD_ADMIN: 'ADD_ADMIN',
  ADD_PRODUCT: 'ADD_PRODUCT',
  BUY_PRODUCTS: 'BUY_PRODUCTS',
  CART_ITEM_ADDED: 'CART_ITEM_ADDED',
  CHANGE_NAME: 'CHANGE_NAME',
  CHANGE_EMAIL: 'CHANGE_EMAIL',
  CHANGE_USERNAME: 'CHANGE_USERNAME',
  CHANGE_PHONE_NUMBER: 'CHANGE_PHONE_NUMBER',
  CHANGE_ADDRESS: 'CHANGE_ADDRESS',
  CHANGE_PASSWORD: 'CHANGE_PASSWORD',
  CHANGE_CONTACTS: 'CHANGE_CONTACTS',
  CHANGE_DELIVERY_ADDRESS: 'CHANGE_DELIVERY_ADDRESS',
  CLEAR_CART: 'CLEAR_CART',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  LOGIN: 'LOGIN',
  REMOVE_ADMIN: 'REMOVE_ADMIN',
  RESET_PASSWORD: 'RESET_PASSWORD',
  PENDING_USER_INFO: 'PENDING_USER_INFO',
  SEND_VERIFICATION_LINK: 'SEND_VERIFICATION_LINK',
  SIGNUP: 'SIGNUP',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: `Price - ascending` },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: `Price - ascending` },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const updateCartActions = {
  INCREMENT: 'INCREMENT',
  DECREMENT: 'DECREMENT',
  NUMBER: 'NUMBER',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};

export const userStatuses = {
  ACTIVE: 'active',
  PENDING: 'pending',
};
