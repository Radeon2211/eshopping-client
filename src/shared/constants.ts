import * as Yup from 'yup';
import countryCodes from 'country-codes-list';
import { ProductCondition, ProductPhotoFieldValue } from './types/enums';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([ProductCondition.NEW, ProductCondition.USED, ProductCondition.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
  photo: Yup.mixed()
    .test('is-error', (value) => value !== ProductPhotoFieldValue.ERROR)
    .nullable(),
};

export const listOfAreaCodes = countryCodes.customArray({
  value: '{countryCallingCode}',
  label: '+{countryCallingCode} {countryNameEn}',
});

export const listOfCountries = countryCodes.customArray({
  value: '{countryNameEn}',
  label: '{countryNameEn}',
});

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: 'Price - ascending' },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: 'Price - ascending' },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig: ScrollToOptions = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};
