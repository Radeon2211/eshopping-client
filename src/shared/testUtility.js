// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
