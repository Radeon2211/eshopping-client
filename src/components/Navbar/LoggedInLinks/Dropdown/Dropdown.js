import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import OutsideClickHandler from 'react-outside-click-handler';
import { useDispatch } from 'react-redux';
import { AnimatePresence } from 'framer-motion';
import * as actions from '../../../../store/actions/indexActions';
import { modalTypes } from '../../../../shared/constants';
import * as SC from './Dropdown.sc';
import { dropdownVariants } from '../../../../shared/framer';
import MyIcon from '../../../UI/MyIcon';
import { ReactComponent as AddProductIcon } from '../../../../images/SVG/add-product.svg';
import { ReactComponent as MyAccountIcon } from '../../../../images/SVG/my-account.svg';
import { ReactComponent as MyOffersIcon } from '../../../../images/SVG/my-offers.svg';
import { ReactComponent as PlacedOrdersIcon } from '../../../../images/SVG/placed-orders.svg';
import { ReactComponent as SellHistoryIcon } from '../../../../images/SVG/sell-history.svg';
import { ReactComponent as LogOutIcon } from '../../../../images/SVG/log-out.svg';

const Dropdown = (props) => {
  const { isVisible, closed } = props;

  const dispatch = useDispatch();
  const onSetModal = useCallback(
    (isModalOpen, modalContent) => dispatch(actions.setModal(isModalOpen, modalContent)),
    [dispatch],
  );

  return (
    <OutsideClickHandler
      onOutsideClick={(e) => {
        if (isVisible && !e.target.closest('#user')) {
          closed();
        }
      }}
    >
      <AnimatePresence exitBeforeEnter>
        {isVisible && (
          <SC.Wrapper variants={dropdownVariants} initial="hidden" animate="visible" exit="hidden">
            <ul className="list">
              <li className="item">
                <div
                  className="link"
                  onClick={() => onSetModal(true, modalTypes.ADD_PRODUCT)}
                  onKeyDown={() => onSetModal(true, modalTypes.ADD_PRODUCT)}
                  role="button"
                  tabIndex={0}
                >
                  <MyIcon size="small">
                    <AddProductIcon />
                  </MyIcon>
                  <span>Add product</span>
                </div>
              </li>
              <li className="item">
                <Link to="/my-account" className="link">
                  <MyIcon size="small">
                    <MyAccountIcon />
                  </MyIcon>
                  <span>My account</span>
                </Link>
              </li>
              <li className="item">
                <Link to="/my-account/products" className="link">
                  <MyIcon size="small">
                    <MyOffersIcon />
                  </MyIcon>
                  <span>My offers</span>
                </Link>
              </li>
              <li className="item">
                <Link to="/my-account/placed-orders" className="link">
                  <MyIcon size="small">
                    <PlacedOrdersIcon />
                  </MyIcon>
                  <span>Placed orders</span>
                </Link>
              </li>
              <li className="item">
                <Link to="/my-account/sell-history" className="link">
                  <MyIcon size="small">
                    <SellHistoryIcon />
                  </MyIcon>
                  <span>My sell history</span>
                </Link>
              </li>
              <li className="item">
                <Link to="/logout" className="link">
                  <MyIcon size="small">
                    <LogOutIcon />
                  </MyIcon>
                  <span>Log out</span>
                </Link>
              </li>
            </ul>
          </SC.Wrapper>
        )}
      </AnimatePresence>
    </OutsideClickHandler>
  );
};

Dropdown.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  closed: PropTypes.func.isRequired,
};

export default Dropdown;
