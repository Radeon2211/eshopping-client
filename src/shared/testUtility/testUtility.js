// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
