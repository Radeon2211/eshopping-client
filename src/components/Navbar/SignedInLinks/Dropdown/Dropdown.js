import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import OutsideClickHandler from 'react-outside-click-handler';
import * as SC from './Dropdown.sc';
import MyIcon from '../../../UI/MyIcon/MyIcon';
import { ReactComponent as AddProductIcon } from '../../../../images/SVG/add-product.svg';
import { ReactComponent as MyAccountIcon } from '../../../../images/SVG/my-account.svg';
import { ReactComponent as MyOffersIcon } from '../../../../images/SVG/my-offers.svg';
import { ReactComponent as PlacedOrdersIcon } from '../../../../images/SVG/placed-orders.svg';
import { ReactComponent as SellHistoryIcon } from '../../../../images/SVG/sell-history.svg';
import { ReactComponent as LogOutIcon } from '../../../../images/SVG/log-out.svg';

const Dropdown = (props) => {
  const { visible, closed } = props;
  return (
    <OutsideClickHandler
      onOutsideClick={(e) => {
        if (visible && !e.target.closest('#user')) closed();
      }}
    >
      <SC.Wrapper visible={visible}>
        <ul className="list">
          <li className="item">
            <Link to="/add-product" className="link">
              <MyIcon size="small"><AddProductIcon /></MyIcon>
              <span>Add product</span>
            </Link>
          </li>
          <li className="item">
            <Link to="/my-account" className="link">
              <MyIcon size="small"><MyAccountIcon /></MyIcon>
              <span>My account</span>
            </Link>
          </li>
          <li className="item">
            <Link to="/my-account/products" className="link">
              <MyIcon size="small"><MyOffersIcon /></MyIcon>
              <span>My offers</span>
            </Link>
          </li>
          <li className="item">
            <Link to="/my-account/placed-orders" className="link">
              <MyIcon size="small"><PlacedOrdersIcon /></MyIcon>
              <span>Placed orders</span>
            </Link>
          </li>
          <li className="item">
            <Link to="/my-account/sell-history" className="link">
              <MyIcon size="small"><SellHistoryIcon /></MyIcon>
              <span>My sell history</span>
            </Link>
          </li>
          <li className="item">
            <Link to="/logout" className="link">
              <MyIcon size="small"><LogOutIcon /></MyIcon>
              <span>Log out</span>
            </Link>
          </li>
        </ul>
      </SC.Wrapper>
    </OutsideClickHandler>
  );
};

Dropdown.propTypes = {
  visible: PropTypes.bool.isRequired,
  closed: PropTypes.func.isRequired,
};

export default Dropdown;
