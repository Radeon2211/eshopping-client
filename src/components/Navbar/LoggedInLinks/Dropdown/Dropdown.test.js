import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import userEvent from '@testing-library/user-event';
import theme from '../../../../styled/theme';
import Dropdown from './Dropdown';
import { checkProps } from '../../../../shared/testUtility/testUtility';
import { modalTypes } from '../../../../shared/constants';
import * as actions from '../../../../store/actions/indexActions';

const mockStore = configureMockStore([thunk]);

const setUp = (isVisible, closed = jest.fn()) => {
  const history = {
    listen: jest.fn(),
    createHref: jest.fn(),
    location: { pathname: '/products', search: '?p=1' },
    push: jest.fn(),
  };

  const store = mockStore({});
  store.dispatch = jest.fn();

  return {
    ...render(
      <Provider store={store}>
        <Router history={history}>
          <ThemeProvider theme={theme}>
            <Dropdown isVisible={isVisible} closed={closed} />
          </ThemeProvider>
        </Router>
      </Provider>,
    ),
    store,
    history,
  };
};

afterEach(cleanup);

describe('<Dropdown />', () => {
  describe('check prop types', () => {
    it('should NOT throw a warning', () => {
      const props = {
        isVisible: true,
        closed: jest.fn(),
      };
      expect(checkProps(Dropdown, props)).toBeUndefined();
    });

    it('should throw a warning', () => {
      expect(checkProps(Dropdown, {})).not.toBe(null);
    });
  });

  describe('check how renders', () => {
    it('should render everything correctly', () => {
      const { asFragment } = setUp(true);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should NOT render anything', () => {
      const { asFragment } = setUp(false);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('check behaviour after clicking at options', () => {
    it('should call push with correct paths after clicking at links', () => {
      const { history } = setUp(true, jest.fn());

      fireEvent.click(screen.getByText(/my account/i));
      expect(history.push).toHaveBeenCalledWith('/my-account/data');

      fireEvent.click(screen.getByText(/my offers/i));
      expect(history.push).toHaveBeenLastCalledWith('/my-account/products?p=1');

      fireEvent.click(screen.getByText(/my sell history/i));
      expect(history.push).toHaveBeenLastCalledWith('/my-account/sell-history?p=1');

      fireEvent.click(screen.getByText(/placed orders/i));
      expect(history.push).toHaveBeenLastCalledWith('/my-account/placed-orders?p=1');

      fireEvent.click(screen.getByText(/log out/i));
      expect(history.push).toHaveBeenLastCalledWith('/logout');

      expect(history.push).toHaveBeenCalledTimes(5);
    });

    it('should call setModal() after add product option click', () => {
      const { store, history } = setUp(true, jest.fn());
      fireEvent.click(screen.getByText(/add product/i));
      expect(store.dispatch).toHaveBeenCalledWith(actions.setModal(modalTypes.ADD_PRODUCT));
      expect(history.push).not.toHaveBeenCalled();
    });
  });

  describe('check closing mechanism', () => {
    it('should call closed after clicking outside if isVisible is true', () => {
      const closedFn = jest.fn();
      setUp(true, closedFn);

      userEvent.click(document.body);
      expect(closedFn).toHaveBeenCalledTimes(1);
    });

    it('should NOT call closed after clicking outside if isVisible is false', () => {
      const closedFn = jest.fn();
      setUp(false, closedFn);

      userEvent.click(document.body);
      expect(closedFn).not.toHaveBeenCalled();
    });

    it('should NOT call closed after clicking at <Dropdown />', () => {
      const closedFn = jest.fn();
      setUp(true, closedFn);

      userEvent.click(screen.getByTestId('Dropdown'));
      expect(closedFn).not.toHaveBeenCalled();
    });
  });
});
