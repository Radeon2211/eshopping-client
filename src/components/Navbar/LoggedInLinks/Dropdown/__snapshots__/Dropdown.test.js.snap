// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> Check how renders Should NOT render anything 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`<Dropdown /> Check how renders Should render everything correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-bdfBwQ gGAgfH"
      data-testid="Dropdown"
      style="opacity: 0; pointer-events: none; transform: none;"
    >
      <ul
        class="list"
      >
        <li
          class="item"
        >
          <div
            class="link"
            role="button"
            tabindex="0"
          >
            <div
              class="sc-gsTCUz cephmd"
              color=""
            >
              <svg>
                add-product.svg
              </svg>
            </div>
            <span>
              Add product
            </span>
          </div>
        </li>
        <li
          class="item"
        >
          <a
            class="link"
            data-testid="Dropdown-my-account-data-link"
          >
            <div
              class="sc-gsTCUz cephmd"
              color=""
            >
              <svg>
                my-account.svg
              </svg>
            </div>
            <span>
              My account
            </span>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class="link"
            data-testid="Dropdown-my-account-products-link"
          >
            <div
              class="sc-gsTCUz cephmd"
              color=""
            >
              <svg>
                my-offers.svg
              </svg>
            </div>
            <span>
              My offers
            </span>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class="link"
            data-testid="Dropdown-my-account-sell-history-link"
          >
            <div
              class="sc-gsTCUz cephmd"
              color=""
            >
              <svg>
                sell-history.svg
              </svg>
            </div>
            <span>
              My sell history
            </span>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class="link"
            data-testid="Dropdown-my-account-placed-orders-link"
          >
            <div
              class="sc-gsTCUz cephmd"
              color=""
            >
              <svg>
                placed-orders.svg
              </svg>
            </div>
            <span>
              Placed orders
            </span>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class="link"
            data-testid="Dropdown-logout-link"
          >
            <div
              class="sc-gsTCUz cephmd"
              color=""
            >
              <svg>
                log-out.svg
              </svg>
            </div>
            <span>
              Log out
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
